### PR TITLE
Infineat Watchface

### PR DESCRIPTION
**Pull request by dmlls**
### This PR implements the new watchface Infineat.
**For anyone interested in trying it, it is already available [here](https://github.com/dmlls/InfiniTime/releases/tag/1.10.0-1-infineat).**
> Note: the linked version differs slightly from the one in this PR, since it adds extended Latin support.

<br>

demo

<br>

This PR also implements a handy `CheckboxList` (check the watchface list in the above demo). I kept this class as general as possible, along the lines of `Tile` or `ApplicationList`, so that it can be used for other purposes as well.

This watchface could be considered "heavy" in terms of flash storage, since it adds two new fonts and an image. However, I've tried to keep things at their absolute minimum. One of the fonts only includes numbers (Bebas Neue, the main font used for the time) and the other (Teko) includes the characters [A-Za-z0-9]. The range of the letters in the Teko font could even be reduced, since they are only used for the days of the week ("Mon", "Tue", "Wed", etc).

The image is the PINE64 logo and has a size of 23x31px. I tried to make its background transparent so that it could be used in other places, but I didn't get it to render correctly. So in the end the background is white. If someone wants to try to make it transparent, I can provide the original SVG.

For a comparison, the [latest build](https://github.com/InfiniTimeOrg/InfiniTime/actions/runs/1930222342) in `develop`:

```
 Memory region         Used Size  Region Size  %age Used
           FLASH:      403028 B       480 KB     82.00%
             RAM:       54872 B        64 KB     83.73%
```

And the Infineat build:

```
Memory region         Used Size  Region Size  %age Used
           FLASH:      416464 B       480 KB     84.73%
             RAM:       54984 B        64 KB     83.90%
```

### CODE REVIEW

**Comment by thefenriswolf** [comment]
Neat

**Comment by ytsejam1138** [comment]
Everytime a pick the Inifineat watchface my PineTime reboots.

**Comment by dmlls** [comment]
@ytsejam1138 mmh interesting... How did you install it exactly? Also, just to check, you got it from here right? https://github.com/dmlls/InfiniTime/releases/tag/1.8.0-infineat

Anyone else having the same issue?

**Comment by ytsejam1138** [comment]
> @ytsejam1138 mmh interesting... How did you install it exactly? Also, just to check, you got it from here right? https://github.com/dmlls/InfiniTime/releases/tag/1.8.0-infineat
> 
> Anyone else having the same issue?

I installed it with the latest Gadgetbridge. And yes I downloaded it from GitHub from the link above.

**Comment by dmlls** [comment]
@ytsejam1138 I just tried upgrading from InfiniTime 1.8.0 with Gadgetbridge and everything works fine. Maybe you can try with nRF Connect? And maybe also try downloading the image again, just in case.

If that doesn't work, please open an issue on my repo, we'll continue the conversation there. Let's keep this space for the PR itself :)

**Comment by NeroBurner** [comment]
open issue for the bootloop here: https://github.com/dmlls/InfiniTime/issues/1

**Comment by joseph58tech** [comment]
I tried to install the image and it broke my watch

**Comment by joseph58tech** [comment]
> I tried to install the image and it broke my watch
> First I couldn't change brightness, so I restored and brightness still didn't work, so I downgraded to 1.7.1 and now Bluetooth stopped connecting 

OK I restarted my watch wich restored back to 1.8.0 and now everything works again

**Comment by NeroBurner** [comment]
@dmlls I've got a quote from @kieranc 
> [M] <neroburner> When is it required/recommended to increase the version number in the settings-class? Every time a new setting is added? Or only if a setting is changed/removed?
> [T] <KieranC> when a new setting is added afaik
> [T] <KieranC> when i added new settings without incrementing it, the initial values were initialised randomly
> [T] <KieranC> so if you increment it, on boot all settings will have their values initialised from... settings.h values? but if it's not incremented they won't

So I think incrementing the version number in `Settings.h` may fix the freeze/reboot error

**Comment by kieranc** [comment]
> The image is the PINE64 logo and has a size of 23x31px. I tried to make its background transparent so that it could be used in other places, but I didn't get it to render correctly. So in the end the background is white. If someone wants to try to make it transparent, I can provide the original SVG.
Nice looking watchface!
As far as the image transparency you might want to look at this option from lv_conf:
```
/*Images pixels with this color will not be drawn (with chroma keying)*/
#define LV_COLOR_TRANSP    LV_COLOR_MAKE(0x6c, 0xFc, 0x6a)         /*LV_COLOR_LIME: pure green*/
```

**Comment by NeroBurner** [comment]
Just tried 11b81157 on InfiniSim and on my dev-device and it works :tada:

**Comment by JF002** [comment]
@dmlls This watchface looks really good, thanks for creating this PR! :1st_place_medal: 

However, as you may know, we are a bit tight in memory, and I try to avoid adding new "big" resources like fonts for now (until we are able to move a few of them to the external memory). Have you already tested your Infineat with existing fonts? It probably don't look as neat as it is with new fonts, but would that be an acceptable compromise?

File: src/components/datetime/DateTimeController.cpp [hunk 1]
```diff
@@ -120,6 +121,10 @@ const char* DateTime::MonthShortToStringLow(Months month) {
   return MonthsStringLow[static_cast<uint8_t>(month)];
 }
 
+const char* DateTime::DayOfWeekShortToStringLow() {
```
Comment (NeroBurner) [inline src/components/datetime/DateTimeController.cpp:124]: just as in https://github.com/InfiniTimeOrg/InfiniTime/pull/1039 this function can be const

```suggestion
const char* DateTime::DayOfWeekShortToStringLow() const {
```

File: src/components/datetime/DateTimeController.h [hunk 2]
```diff
@@ -64,6 +64,7 @@ namespace Pinetime {
       const char* MonthShortToString();
       const char* DayOfWeekShortToString();
       static const char* MonthShortToStringLow(Months month);
+      const char* DayOfWeekShortToStringLow();
```
Comment (NeroBurner) [inline src/components/datetime/DateTimeController.h:67]: just as in https://github.com/InfiniTimeOrg/InfiniTime/pull/1039 this function can be `const`

```suggestion
      const char* DayOfWeekShortToStringLow() const;
```

**Review by NeroBurner — CHANGES_REQUESTED** [review]

**Comment by dmlls** [comment]
> @dmlls This watchface looks really good, thanks for creating this PR! 🥇
> 
> However, as you may know, we are a bit tight in memory, and I try to avoid adding new "big" resources like fonts for now (until we are able to move a few of them to the external memory). Have you already tested your Infineat with existing fonts? It probably don't look as neat as it is with new fonts, but would that be an acceptable compromise?

@JF002 totally understand, I'll see what I can do with the existent fonts ;) I'd like to take as few space from other contributors as possible, great things won't stop coming to InfiniTime!

In any case, if anyone still prefers the original watchface, they can always refer to the fork.

**Comment by dmlls** [comment]
@kieranc @NeroBurner thanks for the suggestions guys!

**Comment by dmlls** [comment]
I added a couple of commits that respectively:
- Make the **Pine64 logo background transparent**. I simply used the option "True color with alpha" when converting the image with the [LVGL Image Converter](https://lvgl.io/tools/imageconverter). This however adds an extra 4 KiB to the C file.
- Remove all new fonts included in the original Infineat watchface and **use existing fonts**. Well, I guess it's a little underwhelming but not terrible either. I personally always found the monospaced fonts in InfiniTime looking a bit old-fashioned (except for the Terminal watchface ;).

demo_existing_fonts

The new stats are:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      411524 B       480 KB     83.72%
             RAM:       54520 B        64 KB     83.19%
```

That's 1% less of flash usage with respect to the original Infineat watchface.

**Comment by NeroBurner** [comment]
It still looks neat indeed!

Could you add a battery-symbol to the watchface? I'd like to see the battery level on my watchface without needing to look at either gadgetbridge or the pinetime-settings. Otherwise this would be the first watchface without some sort of battery indicator

**Comment by joseph58tech** [comment]
> It still looks neat indeed!
> 
> Could you add a battery-symbol to the watchface? I'd like to see the battery level on my watchface without needing to look at either gadgetbridge or the pinetime-settings. Otherwise this would be the first watchface without some sort of battery indicator

Yeah I found it weird that there is no battery indicator, I just noticed this about a day ago when I was going to look down at my watch so I could check the battery, and there was no watch face indicator.

**Comment by kieranc** [comment]
As a cleaner option than adding a battery icon you could make the bar on the left go red when battery is below a preset percentage. maybe 40% currently would be safe-ish

**Comment by dmlls** [comment]
@kieranc that's also a good idea. Only thing that it may not that intuitive for users who know nothing about the watchface. I actually had thought of something similar (see https://github.com/dmlls/InfiniTime/pull/3#issuecomment-1072860833).

I will make a couple mockups with different approaches and post them here to gather feedback.

**Comment by dmlls** [comment]
So, this is what I came up with:

**Option 1**: battery icon bottom-left.


icon_bottom_left

<br/>

**Option 2**: battery icon top-left.


icon_top_left

<br/>

**Option 3**: Pine64 logo as battery indicator.


logo_fill



The mockups have the original font, but it would obviously be the same with the system font.
The 3rd option seems to me the less "intrusive" one, although I also like the 1st one.

I'd like to do a small voting with reactions:

- **Option 1**: 🚀
- **Option 2**: 🎉
- **Option 3**: ❤️

**Comment by ytsejam1138** [comment]
Option 3: ❤️ 

But use a darker color on the pinecone to give it more contrast so it will be easier to read at a glance.

**Comment by lman0** [comment]
option 3 :  but as well a little more darker
and in order to see , in action , can you link for each of them (or at least 1 and 3 )a dfu ?

**Comment by lman0** [comment]
it lack as well a state for when charing @dmlls  , maybe maybe the pinecone binking woulf help for optio 3

**Comment by dmlls** [comment]
So... I implemented the battery indicator (option 3) with a nice little charging animation. Here's a demo:


demo_battery


I also rebased upstream.

If you guys could take it for a spin and test that everything works fine, that'd be great! 😀

**Comment by Avamander** [comment]
I let the CI to run, so people can potentially use those builds to test.

**Comment by Zandengoff** [comment]
Really excited for this one. When I got my Pinetime, I found the watch faces by zephyrlabs and was immediately bummed there was no way to load them without integrating them into the firmware. Terminal is cool and I have been enjoying it, but really want to see more fancy watchfaces. This really hits the spot!

**Comment by trman** [comment]
hi 
tested around a week , no problem of usability 
no crash whatosever
and the battery is big enough to be readable and understandable (at very least 40% and 60% are very clear)
if anything , it would be great to have the hr data visible (if it activated) at the right , maybe, of the step counter

**Comment by dmlls** [comment]
@Zandengoff thank you!

@trman thanks too for the testing!

> if anything , it would be great to have the hr data visible (if it activated) at the right , maybe, of the step counter.

This point was discussed briefly here https://github.com/dmlls/InfiniTime/pull/3#issuecomment-1072860833. I'm open to continue this discussion, but in my opinion adding the heart rate information takes more from the cleanliness of the design than functionality it provides. Unfortunately, the PineTime screen is rather small and things get crowded pretty easily ;)

**Comment by lman0** [comment]
hello 

the watchface is quite neat , so i decided to give it a try @dmlls !
after toying around , i found  the background change quite good.
but then i clicked on the bottom " on" label ...
i found that when the background is removed , so is the battery icon !
that is quite inconvenient :(

could you make remain when off , only the little white background and the battery icon ?
maybe another way , would to make the battery icon but with it color inverted
 ( so it could become a white pine battery on black background)

**Comment by dmlls** [comment]
@lman0 that's a very good catch! I'll try your proposed ways and see which one fits better ✌🏻

**Comment by Albirew** [comment]
wow that's neat! and using the pine logo as battery indicator is genius!
quick suggestion: changing color to red when battery < 40% or green when battery > 90%
either the color of the pine only or the theme's "sheets", i dunno witch one would be the best

**Comment by lman0** [comment]
@dmlls  any news ?

**Comment by dmlls** [comment]
@lman0 sorry guys, it's been very busy lately... I finished implementing the battery indicator without the side cover and it's looking good. Now I just have to rebase `develop`. I hope it's done by the end of this weekend!

**Comment by dmlls** [comment]
@Albirew I think changing the color of the battery when its level it's critical (e.g., < 30%) could be a good idea, serving a functional purpose. On the other hand, I'm afraid that changing it also when it's more than > 90 % could be too much color change. I'd like to keep the watchface simple and predictable :)

**Comment by dmlls** [comment]
I rebased `develop` and implemented the battery indicator when the side cover is off:

demo_battery_no_cover_cropped_pr

If anyone wants to get the original watchface with these changes, you can head to https://github.com/dmlls/InfiniTime/releases/tag/1.9.0-2-infineat. That release is based off InfiniTime 1.9.0 stable.

Also, I wanted to mention that I don't know if it makes a lot of sense to keep this PR updated, since it is unlikely it will be merged. Right now I need to keep two different codebases (because of the different fonts), and it is quite a pain.

**Comment by devnoname120** [comment]
@dmlls Why is it unlikely to get merged?

**Comment by trman** [comment]
@devnoname120 because the neat fonts of infinineat  (and  picture icon and latin fonts needed)  make it too big to fit in the firmware right now.
@dmlls need to wait at least until , the fonts are put in the external memory , that don't seem the priority right now

**Comment by JF002** [comment]
@trman 
> but it don't seem the infinitime core dev priority right now

Using the external flash memory to store images and fonts is definitely on the top of my TODO list :) You can see my progress [here](https://github.com/InfiniTimeOrg/InfiniTime/issues/321). As you can see, it's kinda working, but not as good as I would like to (transitions are slower, fonts are loaded in RAM,...). So there's still a bit of work to do.

@dmlls 
> Also, I wanted to mention that I don't know if it makes a lot of sense to keep this PR updated, since it is unlikely it will be merged. Right now I need to keep two different codebases (because of the different fonts), and it is quite a pain.

I understand your pain and I'm really sorry about that! I do hope however that we'll be able to merge your work as soon as possible :)

**Comment by JF002** [comment]
@dmlls I would like to do some test with the Infineat watchface and the external flash memory. Could you provide me with the original picture and font files so I can convert them to a binary file?

**Comment by dmlls** [comment]
@JF002 that's great to hear!

- The font files can be found here: [BebasNeue-Regular.ttf](https://github.com/dmlls/InfiniTime/blob/infineat-color/src/displayapp/fonts/BebasNeue-Regular.ttf) and [Teko-Light.ttf](https://github.com/dmlls/InfiniTime/blob/infineat-color/src/displayapp/fonts/Teko-Light.ttf).
- The logo:
    - [png](https://user-images.githubusercontent.com/22967053/175952822-10dda207-0811-4b7e-8700-39eafc6e538a.png) (already with the appropriate dimensions, i.e., 23x31px).
    - [svg](https://user-images.githubusercontent.com/22967053/175953274-67326262-1a17-4b90-a3b4-daca4afd18a0.svg) (original Inkscape file).

Hope that helps, anything else let me know.

And nice numbers indeed! 😄

**Comment by JF002** [comment]
Thanks @dmlls. Have a look at the video [in this comment](https://github.com/InfiniTimeOrg/InfiniTime/issues/321#issuecomment-1170448754) :)

**Comment by lman0** [comment]
@dmlls  could you update your fork to 1.10.0 ?
i understand that updating this pr might be a pain since it can't be merged
 unless @JF002 study of external flash usage turn into a proper "merged" pr 
 but since it's only the fork , it shouldn't be too hard for you to update the fork
 
 
 after all  there is people , that may would like to use infineat with updated code base too (with latin font)!

**Comment by dmlls** [comment]
> @dmlls could you update your fork to 1.10.0 ?

@lman0 sure, will keep it up to date until the ongoing work on external flash usage is finished.
That having said, this one will need to wait a couple of weeks since I don't have my PineTime at the moment 😅

**Comment by dmlls** [comment]
@lman0 fork updated to 1.10.0, [available here](https://github.com/dmlls/InfiniTime/releases/tag/1.10.0-1-infineat).
Sorry for the delay, it's been very busy lately!

**Comment by lman0** [comment]
thanks @dmlls !

**Comment by JF002** [comment]
Thank you very much for your contribution, @dmlls !
I can finally merge this new watchface now that InfiniTime supports loading resources in the external memory!

**Pull request merged**